artichoke-backend/src/extn/core/time/trampoline.rs

```diff
@@ -1,35 +1,151 @@
+        )
+        .into())
+    }
+}
 
+// Check a Ruby Value Hash for the appropriate values to create and Offset
+// for a _time_.
+fn offset_from_options(interp: &mut Artichoke, options: Value) -> Result<Offset, Error> {
+    let hash: Vec<(Value, Value)> = interp.try_convert_mut(options)?;
+    let tz = hash
+        .into_iter()
+        .map(|(k, v)| {
+            let key = unsafe { Symbol::unbox_from_value(&mut k.clone(), interp)? }.bytes(interp);
+            if key == b"in" {
+                Ok(v)
+            } else {
+                Err(ArgumentError::with_message("unknown keyword"))?
+            }
+        })
+        .collect::<Result<Vec<Value>, Error>>()?;
+
+    match tz[..] {
+        [tz] => offset_from_value(interp, tz),
+        _ => Err(ArgumentError::with_message("unknown keyword"))?,
+    }
+}
+
+// Constructor
 pub fn now(interp: &mut Artichoke) -> Result<Value, Error> {
-    let now = Time::now();
+    let now = Time::now()?;
     let result = Time::alloc_value(now, interp)?;
     Ok(result)
 }
 
-pub fn at(interp: &mut Artichoke, seconds: Value, microseconds: Option<Value>) -> Result<Value, Error> {
+pub fn at(
+    interp: &mut Artichoke,
+    seconds: Value,
+    first: Option<Value>,
+    second: Option<Value>,
+    third: Option<Value>,
+) -> Result<Value, Error> {
+    // Coerce the params to the correct place. Specifically:
+    // - the options hash might not always be provided as the last argument.
+    // - subseconds can be provided with an optional symbol for the type of subsec.
+    //
+    // ```console
+    // [3.1.2] > Time.at(1)
+    // => 1970-01-01 01:00:01 +0100
+    // [3.1.2] > Time.at(1, 1)
+    // => 1970-01-01 01:00:01.000001 +0100
+    // [3.1.2] > Time.at(1, 1, :nsec)
+    // => 1970-01-01 01:00:01.000000001 +0100
+    // [3.1.2] > Time.at(1, in: "A")
+    // => 1970-01-01 01:00:01 +0100
+    // [3.1.2] > Time.at(1, 1, in: "A")
+    // => 1970-01-01 01:00:01.000001 +0100
+    // [3.1.2] > Time.at(1, 1, :nsec)
+    // => 1970-01-01 01:00:01.000000001 +0100
+    // [3.1.2] > Time.at(1, 1, :nsec, in: "A")
+    // => 1970-01-01 01:00:01.000000001 +0100
+    // ```
+    let (subsec, subsec_type, options) = match (
+        first.map(|opt| opt.ruby_type()),
+        second.map(|opt| opt.ruby_type()),
+        third.map(|opt| opt.ruby_type()),
+    ) {
+        (None, None, None) => (None, None, None),
+        (Some(Ruby::Hash), None, None) => (None, None, first),
+        (Some(Ruby::Fixnum), None, None) => (first, None, None),
+        (Some(_), None, None) => Err(ArgumentError::with_message("expected a number"))?,
+        (Some(Ruby::Fixnum), Some(Ruby::Hash), None) => (first, None, second),
+        (Some(Ruby::Fixnum), Some(Ruby::Symbol), None) => (first, second, None),
+        (Some(Ruby::Fixnum), Some(_), None) => Err(ArgumentError::with_message(
+            "expected one of [:milliseconds, :usec, :nsec]",
```

Comment:
we can add this as a todo but it looks like the error message is `"unexpected unit: #{x.inspect}"`:

```
[3.1.2] > Time.at(1, 1, :xxx)
<internal:timev>:274:in `at': unexpected unit: xxx (ArgumentError)
        from (irb):12:in `<main>'
        from /usr/local/var/rbenv/versions/3.1.2/lib/ruby/gems/3.1.0/gems/irb-1.4.1/exe/irb:11:in `<top (required)>'
        from /usr/local/var/rbenv/versions/3.1.2/bin/irb:25:in `load'
        from /usr/local/var/rbenv/versions/3.1.2/bin/irb:25:in `<main>'
[3.1.2] > Time.at(1, 1, Object.new)
<internal:timev>:274:in `at': unexpected unit: #<Object:0x0000000111579b78> (ArgumentError)
        from (irb):13:in `<main>'
        from /usr/local/var/rbenv/versions/3.1.2/lib/ruby/gems/3.1.0/gems/irb-1.4.1/exe/irb:11:in `<top (required)>'
        from /usr/local/var/rbenv/versions/3.1.2/bin/irb:25:in `load'
        from /usr/local/var/rbenv/versions/3.1.2/bin/irb:25:in `<main>'
[3.1.2] > Time.at(1, 1, "nsec")
<internal:timev>:274:in `at': unexpected unit: nsec (ArgumentError)
        from (irb):14:in `<main>'
        from /usr/local/var/rbenv/versions/3.1.2/lib/ruby/gems/3.1.0/gems/irb-1.4.1/exe/irb:11:in `<top (required)>'
        from /usr/local/var/rbenv/versions/3.1.2/bin/irb:25:in `load'
        from /usr/local/var/rbenv/versions/3.1.2/bin/irb:25:in `<main>'
```